Sources/SwiftParser/Expressions.swift

```diff
@@ -2017,6 +2056,37 @@ extension Parser {
 }
 
 extension Parser.Lookahead {
+  /// Check if the keypath method is applied, and not partially applied which should be parsed as a key path property.
+  mutating func isAppliedKeyPathMethod() -> Bool {
+    var lookahead = self.lookahead()
+    var hasLParen = false, hasRParen = false
+
+    while true {
+      let token = lookahead.peek().rawTokenKind
+      if token == .endOfFile || lookahead.atStartOfLine {
+        break
+      }
+      if token == .leftParen {
+        hasLParen = true
+      }
+      if token == .colon {
+        lookahead.consumeAnyToken()
+        // If there's a colon followed by a right parenthesis, it is
+        // a partial application and should be parsed as a property.
+        if lookahead.peek().rawTokenKind == .rightParen {
+          return false
+        }
+      }
+      if token == .rightParen {
+        hasRParen = true
+      }
+      lookahead.consumeAnyToken()
+    }
+    // If parentheses exist with no partial application pattern,
+    // parse as a key path method.
+    return hasLParen && hasRParen ? true : false
```

Comment:
This seems problematic to me. I haven’t fully though it through but it has the potential for an unbounded lookahead, which could be very bad for parser performance and the use of `consumeAnyToken` to eat any uncovered token also seems like a pitfall to me.

What I would suggest is that in your actual `parse*` method, you call `parseDottedExpressionSuffix` outside of the `if`. That would parse any compound names, ie. unapplied function references. If we are at a parenthesis after that, continue parsing arguments for an applied key path reference. That way, you can also parse key paths like `\Foo.t(a:)(2)`, mirroring the following valid Swift code

```swift
struct Foo {
  func t(a: Int) {
    self.t(a:)(2)
  }
}
```

Or am I missing something with that approach?

Reply:
This makes sense and I've updated the code to reflect this. For reasons I can't reproduce now, `parseDottedExpressionSuffix` did not work for me initially and so I wrote this new method. Regardless it works now. All good things to know about `Lookahead` as well. Thank you!